### ISSUE TITLE
Failed to initialize video producer. Error code: 4099

### ISSUE THREAD

**Issue by Gil401** [aws-robotics/kinesisvideo-ros1#2]
Hi Guys,

I have followed the readme, however when tried to roslaunch the kinesis_video_streamer.launch I am getting the following error:
`[FATAL] [1544446273.975483790]: [Initialize] Failed to initialize video producer. Error code: 4099
`
I have updated the yaml files to have to correct topic (sensor_msgs::Image), correct topic type(2) and correct region (eu-central-1)

What else am I missing?

Thanks

**Comment by juanrh** [comment]
Hi, 

Thanks for taking a look to our kinesis video streamer node. 

That error usually happens when the AWS credentials are not configured, and the node is not able to initialize the kinesis video client. Can you please check that the credentials and resource permissions have are set up according to [these instructions](https://github.com/aws-robotics/kinesisvideo-ros1#aws-credentials). Please let me know how it went!

**Comment by Gil401** [comment]
Hi @juanrh  
thanks for assisting with this. However it appears that everything is set as required (IAM permissions and credentials in ~/.aws/credentials and ~/.aws/config)

**Comment by juanrh** [comment]
Hi Gil, 

Thanks for your patience. Error code 4099 / 0x1003 corresponds to [KINESIS_MANAGER_STATUS_INVALID_INPUT](https://github.com/aws-robotics/kinesisvideo-common/blob/master/kinesis_manager/include/kinesis_manager/common.h#L34) and it is thrown during the video producer initialization [when there is some configuration problem](https://github.com/aws-robotics/kinesisvideo-common/blob/master/kinesis_manager/src/kinesis_stream_manager.cpp#L189). I think in this case there might be a bug in our code because even though the [sample configuration includes the region parameter](https://github.com/aws-robotics/kinesisvideo-ros1/blob/master/kinesis_video_streamer/config/sample_configuration.yaml#L30) it is not always read correctly. 

A possible _workaround_ would be setting a system wide default AWS region. For that you should create or edit the file `~/.aws/config` to add the following lines:

```
[default]
region=us-west-2
```

Could you please also provide the `PARAMETERS` section printed by the `roslaunch` command, so we are sure the region is passed? There should be an entry `/kinesis_video_streamer/kinesis_video/stream0/aws_client_configuration/region`.

I'll also work on a permanent code fix, but I also would like to try to unblock you as soon as possible. 

Thanks

**Comment by juanrh** [comment]
Hi Gil, 

What I was asking is that you added the output printed by the `roslaunch` command to this email thread, in particular the `PARAMETERS` section. 

Thanks

**Comment by Gil401** [comment]
```
SUMMARY
========

PARAMETERS
 * /kinesis_video_streamer/kinesis_video/log4cplus_config: /my_workspace/v...
 * /kinesis_video_streamer/kinesis_video/stream_count: 1
 * /rosdistro: kinetic
 * /rosversion: 1.12.14

NODES
  /
    kinesis_video_streamer (kinesis_video_streamer/kinesis_video_streamer)

```

**Comment by juanrh** [comment]
Hi Gil, 

I was expecting more configuration parameters, and also 

```
NODES
  /
    h264_video_encoder (h264_video_encoder/h264_video_encoder)
    kinesis_video_streamer (kinesis_video_streamer/kinesis_video_streamer)
```

How are you launching the node? Are you using `roslaunch kinesis_video_streamer sample_application.launch`?

**Comment by Gil401** [comment]
Here with` roslaunch kinesis_video_streamer sample_application.launch`
SUMMARY
========

PARAMETERS
 * /h264_video_encoder/bitrate: 2048000
 * /h264_video_encoder/fps_denominator: 1
 * /h264_video_encoder/fps_numerator: 30
 * /h264_video_encoder/image_transport: compressed
 * /h264_video_encoder/publication_topic: /video/encoded
 * /h264_video_encoder/subscription_topic: /camera/raw
 * /kinesis_video_streamer/kinesis_video/log4cplus_config: /my_workspace/v...
 * /kinesis_video_streamer/kinesis_video/stream0/aws_client_configuration/connect_timeout_ms: 9000
 * /kinesis_video_streamer/kinesis_video/stream0/aws_client_configuration/max_retries: 10
 * /kinesis_video_streamer/kinesis_video/stream0/aws_client_configuration/region: eu-central-1
 * /kinesis_video_streamer/kinesis_video/stream0/aws_client_configuration/request_timeout_ms: 9000
 * /kinesis_video_streamer/kinesis_video/stream0/frame_rate: 10
 * /kinesis_video_streamer/kinesis_video/stream0/frame_timecodes: False
 * /kinesis_video_streamer/kinesis_video/stream0/stream_name: my_stream1
 * /kinesis_video_streamer/kinesis_video/stream0/subscription_topic: /camera/36...
 * /kinesis_video_streamer/kinesis_video/stream0/topic_type: 2
 * /kinesis_video_streamer/kinesis_video/stream_count: 1
 * /rosdistro: kinetic
 * /rosversion: 1.12.14

NODES
  /
    h264_video_encoder (h264_video_encoder/h264_video_encoder)
    kinesis_video_streamer (kinesis_video_streamer/kinesis_video_streamer)

**Comment by juanrh** [comment]
Hi Gil, 

I sent a new PR to fix a configuration issue that resulted in the AWS configuration being ignored. You could try the change using the [branch `node-stream-launch-files` of my fork of kinesisvideo-ros1](https://github.com/juanrh/kinesisvideo-ros1/tree/node-stream-launch-files). However, it might be the case that this bug wasn't affecting you, because you already had the region configured on `~/.aws/config`. Can you please add the full output printed by the `roslaunch kinesis_video_streamer sample_application.launch` command to this thread? That way we could get more context about was is failing. 

Thanks again for your patience

**Comment by Gil401** [comment]
Hi @juanrh thank you for the assistance, however, even with your version I am still getting the same error

here is the output of `$ roslaunch kinesis_video_streamer sample_application.launch `

```
started roslaunch server http://127.0.0.1:38939/

SUMMARY
========

PARAMETERS
 * /h264_video_encoder/bitrate: 2048000
 * /h264_video_encodכer/fps_denominator: 1
 * /h264_video_encoder/fps_numerator: 30
 * /h264_video_encoder/image_transport: compressed
 * /h264_video_encoder/publication_topic: /video/encoded
 * /h264_video_encoder/subscription_topic: /camera/36...
 * /kinesis_video_streamer/kinesis_video/log4cplus_config: /my_workspace/v...
 * /kinesis_video_streamer/kinesis_video/stream0/aws_client_configuration/connect_timeout_ms: 9000
 * /kinesis_video_streamer/kinesis_video/stream0/aws_client_configuration/max_retries: 10
 * /kinesis_video_streamer/kinesis_video/stream0/aws_client_configuration/region: eu-central-1
 * /kinesis_video_streamer/kinesis_video/stream0/aws_client_configuration/request_timeout_ms: 9000
 * /kinesis_video_streamer/kinesis_video/stream0/frame_rate: 10
 * /kinesis_video_streamer/kinesis_video/stream0/frame_timecodes: False
 * /kinesis_video_streamer/kinesis_video/stream0/stream_name: my_stream1
 * /kinesis_video_streamer/kinesis_video/stream0/subscription_topic: /camera/36...
 * /kinesis_video_streamer/kinesis_video/stream0/topic_type: 2
 * /kinesis_video_streamer/kinesis_video/stream_count: 1
 * /rosdistro: kinetic
 * /rosversion: 1.12.14

NODES
  /
    h264_video_encoder (h264_video_encoder/h264_video_encoder)
    kinesis_video_streamer (kinesis_video_streamer/kinesis_video_streamer)

ROS_MASTER_URI=http://localhost:11311

process[kinesis_video_streamer-1]: started with pid [22544]
process[h264_video_encoder-2]: started with pid [22550]
[FATAL] [1544594903.235860450]: [Initialize] Failed to initialize video producer. Error code: 4099
[kinesis_video_streamer-1] process has died [pid 22544, exit code 3, cmd /my_workspace/videoStream/kvs_ws/install/kinesis_video_streamer/lib/kinesis_video_streamer/kinesis_video_streamer __name:=kinesis_video_streamer __log:=/home/user/.ros/log/fe706f1c-fd4f-11e8-a92e-e04f4393db72/kinesis_video_streamer-1.log].
log file: /home/user/.ros/log/fe706f1c-fd4f-11e8-a92e-e04f4393db72/kinesis_video_streamer-1*.log
^C[h264_video_encoder-2] killing on exit
shutting down processing monitor...
... shutting down processing monitor complete
```

**Comment by juanrh** [comment]
Hi Gil,

It would help us a lot if we could see some additional log files. It would be nice if you could edit ` $ROS_ROOT/config/rosconsole.config` (usually this points to /opt/ros/kinetic/share/ros/config/rosconsole.config) to set the log level to debug, e.g. as follows 

```
#log4j.logger.ros=INFO
log4j.logger.ros=DEBUG
```

and then run `roslaunch kinesis_video_streamer sample_application.launch` again, and attach to this issue a compressed file for the log path shown in the standard output (it should be something like `~/.ros/log/e97c3c54-fe66-11e8-9f31-7c76355c5e9f`)

Thanks

**Comment by Gil401** [comment]
[error.txt](https://github.com/aws-robotics/kinesisvideo-ros1/files/2674683/error.txt)

**Comment by juanrh** [comment]
Hi, 

Thanks a lot for the information. Usually more than 1 log file is produced, for example I get

```
h264_video_encoder-3-stdout.log
kinesis_video_streamer-2-stdout.log
master.log
roslaunch-ub80c134adc2254.log
rosout-1-stdout.log
rosout.log
```

are you only getting a single log file?, did you make any additional change the log configuration in ROS? If you have more log files it would be nice if you could attach them here. 

Thanks again for your patience

**Comment by Gil401** [comment]
Yey!! It is working!!!! Your branch fixed it!

Thanks a lot for the assistance!!
Highly appreciated!!

**Comment by juanrh** [comment]
I was my pleasure! Thanks again for taking a look to our kinesis video streamer node, and helping us find that bug. I'll take care of ensuring the fix is merged to master as soon as possible.

**Comment by juanrh** [comment]
Hi @Gil401, just a heads up that this fix was already merged to the master branch. 

Thanks,

Juan

**Comment by Gil401** [comment]
Hi @juanrh 

Long time :)

I have used `sudo apt-get install ros-kinetic-kinesis-video-streamer ` and trying to run the node again, however the same error msg pops up.

This time it looks like it is unable to pull the aws config since this is what I am getting in the logs:

> [32m[DEBUG] [1550416833.665917198]: XML-RPC call [searchParam] returned an error (-1): [Cannot find parameter [aws_client_configuration/region] in an upwards search][0m
> [32m[DEBUG] [1550416833.666718549]: XML-RPC call [searchParam] returned an error (-1): [Cannot find parameter [aws_client_configuration/user_agent] in an upwards search][0m
> [32m[DEBUG] [1550416833.667196506]: XML-RPC call [searchParam] returned an error (-1): [Cannot find parameter [aws_client_configuration/endpoint_override] in an upwards search][0m
> [32m[DEBUG] [1550416833.667942589]: XML-RPC call [searchParam] returned an error (-1): [Cannot find parameter [aws_client_configuration/proxy_host] in an upwards search][0m
> [32m[DEBUG] [1550416833.668966060]: XML-RPC call [searchParam] returned an error (-1): [Cannot find parameter [aws_client_configuration/proxy_user_name] in an upwards search][0m
> [32m[DEBUG] [1550416833.669568456]: XML-RPC call [searchParam] returned an error (-1): [Cannot find parameter [aws_client_configuration/proxy_password] in an upwards search][0m
> [32m[DEBUG] [1550416833.670347282]: XML-RPC call [searchParam] returned an error (-1): [Cannot find parameter [aws_client_configuration/ca_path] in an upwards search][0m
> [32m[DEBUG] [1550416833.671793340]: XML-RPC call [searchParam] returned an error (-1): [Cannot find parameter [aws_client_configuration/ca_file] in an upwards search][0m
> [32m[DEBUG] [1550416833.673628852]: XML-RPC call [searchParam] returned an error (-1): [Cannot find parameter [aws_client_configuration/request_timeout_ms] in an upwards search][0m
> [32m[DEBUG] [1550416833.674503187]: XML-RPC call [searchParam] returned an error (-1): [Cannot find parameter [aws_client_configuration/connect_timeout_ms] in an upwards search][0m
> [32m[DEBUG] [1550416833.675344114]: XML-RPC call [searchParam] returned an error (-1): [Cannot find parameter [aws_client_configuration/max_connections] in an upwards search][0m
> [32m[DEBUG] [1550416833.676414532]: XML-RPC call [searchParam] returned an error (-1): [Cannot find parameter [aws_client_configuration/proxy_port] in an upwards search][0m
> [32m[DEBUG] [1550416833.677388551]: XML-RPC call [searchParam] returned an error (-1): [Cannot find parameter [aws_client_configuration/use_dual_stack] in an upwards search][0m
> [32m[DEBUG] [1550416833.678286534]: XML-RPC call [searchParam] returned an error (-1): [Cannot find parameter [aws_client_configuration/enable_clock_skew_adjustment] in an upwards search][0m
> [32m[DEBUG] [1550416833.679101957]: XML-RPC call [searchParam] returned an error (-1): [Cannot find parameter [aws_client_configuration/follow_redirects] in an upwards search][0m
> [32m[DEBUG] [1550416833.679574468]: XML-RPC call [searchParam] returned an error (-1): [Cannot find parameter [aws_client_configuration/verify_SSL] in an upwards search][0m
> [32m[DEBUG] [1550416833.680008093]: XML-RPC call [searchParam] returned an error (-1): [Cannot find parameter [aws_client_configuration/max_retries] in an upwards search][0m

even though they are set in sample_configuration.yaml

Any idea? :)

Thanks again

**Comment by AAlon** [comment]
Mostly identical to #16 so closing this and will track the apt sync in #16.
